base_control/Movement.cpp

```diff
@@ -0,0 +1,315 @@
+  front_right_motor_.backward();
+  back_right_motor_.backward();
+}
+
+void Movement::move315() {
+  front_left_motor_.stop();
+  back_left_motor_.backward();
+  front_right_motor_.backward();
+  back_right_motor_.stop();
+}
+
+void Movement::rotateLeft() {
+  front_left_motor_.backward();
+  back_left_motor_.backward();
+  front_right_motor_.forward();
+  back_right_motor_.forward();  
+}
+
+void Movement::rotateRight() {
+  front_left_motor_.forward();
+  back_left_motor_.forward();
+  front_right_motor_.backward();
+  back_right_motor_.backward();
+}
+
+//////////////////////////////////PID//////////////////////////////////////
+void Movement::constantLinearSpeed() {
+  front_right_motor_.constantSpeed(getTargetLinearVelocity());
+  front_left_motor_.constantSpeed(getTargetLinearVelocity());
+  back_left_motor_.constantSpeed(getTargetLinearVelocity());
+  back_right_motor_.constantSpeed(getTargetLinearVelocity());
+}
+
+void Movement::constantAngularSpeed() {
+  front_right_motor_.constantSpeed(getTargetAngularVelocity());
+  front_left_motor_.constantSpeed(getTargetAngularVelocity());
+  back_left_motor_.constantSpeed(getTargetAngularVelocity());
+  back_right_motor_.constantSpeed(getTargetAngularVelocity());
+}
+
+void Movement::velocityAdjustment(const int adjustment) {
+  back_left_motor_.setVelocityAdjustment(
+    (back_left_motor_.getCurrentState() == MotorState::Forward) ? adjustment * -1 : adjustment
+  );
+  front_left_motor_.setVelocityAdjustment(
+    (front_left_motor_.getCurrentState() == MotorState::Forward) ? adjustment * -1 : adjustment
+  );
+  back_right_motor_.setVelocityAdjustment(
+    (back_right_motor_.getCurrentState() == MotorState::Backward) ? adjustment * -1 : adjustment
+  );
+  front_right_motor_.setVelocityAdjustment(
+    (front_right_motor_.getCurrentState() == MotorState::Backward) ? adjustment * -1 : adjustment
+  );
+}
+
+void Movement::pidLinearMovement() {
+    setDirection(angleToDirection(getTargetAngle()));
+    constantLinearSpeed();
+
+    double angle_error = 0;
+    Direction where = whereToGo(angle_error);
+    pid_straight_.compute(angle_error,  straight_output_, 0);
+    velocityAdjustment(straight_output_);
+}
+
+void Movement::pidAngularMovement() {
+    if(delta_angular_<0) {
+        rotateRight();
+    } else {
+        rotateLeft();
+    }
+    constantAngularSpeed();
+}
+
+bool Movement::pidRotate(const double target_angle) {
+    double output = 0;
+    double angle_error = 0;
+    Direction where = whereToGo(angle_error, target_angle);
+    
+    if(abs(angle_error) <= kPidRotationTolerance && 
+       abs(pid_rotation_.getPre()) <= kPidRotationTolerance && 
+       angle_error * pid_rotation_.getPre() >= 0) {
+       
+        stop();
+        return true;
+    }
+
+    pid_rotation_.compute(angle_error,  output, 1);
+    output = abs(output)+kOutputAdjustment;
+    
+    if(where == Direction::left) {
+        rotateRight();
+    } else {
+        rotateLeft();
+    }
```

Comment:
I'm no expert, but why if the Direction is left do we rotate to the right?

Reply:
its true whereToGo was returning the oppostide side.